File: C++/14-Backtracking/Problems/3-CombinationSum/CombinationSumPro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>

using namespace std;

// 1. Backtracking with Duplicates Allowed (Basic Approach)
//   - Allows repeated use of numbers from the candidates array.
//   - Directly explores all possible combinations.
//   - Suitable for problems where the same element can be chosen multiple times.
//   - Example: Finding all possible combinations of coins to make a target sum.
void combinationSum1(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
    if (target == 0) {
        result.push_back(combination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        combination.push_back(candidates[i]);
        combinationSum1(candidates, target - candidates[i], result, combination, i); // Important: i, not i+1
        combination.pop_back();
    }
}

// 2. Backtracking with Duplicates Allowed (Optimized)
//   - Similar to the first approach, but with slight optimization.
//   - Avoids unnecessary recursive calls by checking if the target becomes negative.
void combinationSum2(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
     if (target == 0) {
        result.push_back(combination);
        return;
    }

    for (int i = start; i < candidates.size(); ++i) {
        if (target - candidates[i] >= 0) { // Optimization: Check before recursion
            combination.push_back(candidates[i]);
            combinationSum2(candidates, target - candidates[i], result, combination, i);
            combination.pop_back();
        }
    }
}


// 3. Backtracking with Duplicates Not Allowed
//   - Each number in the candidates array can be used only once in a combination.
//   - Requires sorting the candidates array to handle duplicates effectively.
//   - Example: Finding unique subsets of a set that sum to a target.
void combinationSum3(vector<int>& candidates, int target, vector<vector<int>>& result, vector<int>& combination, int start) {
    if (target == 0) {
        result.push_back(combination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        // Skip duplicate numbers to avoid duplicate combinations
        if (i > start && candidates[i] == candidates[i - 1]) {
            continue;
        }
        combination.push_back(candidates[i]);
        combinationSum3(candidates, target - candidates[i], result, combination, i + 1); // Important: i+1
        combination.pop_back();
    }
}

// 4. Using set to avoid Duplicates
//  - Uses a set to store the results, thus removing any duplicate combinations.
void combinationSum4(vector<int>& candidates, int target, set<vector<int>>& result_set, vector<int>& combination, int start) {
    if (target == 0) {
        // Sort the combination to handle duplicates correctly in the set.
        vector<int> sortedCombination = combination;
        sort(sortedCombination.begin(), sortedCombination.end());
        result_set.insert(sortedCombination);
        return;
    }
    if (target < 0) {
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        combination.push_back(candidates[i]);
        combinationSum4(candidates, target - candidates[i], result_set, combination, i);
        combination.pop_back();
    }

}

// 5. Iterative Approach using Stack
// - Iterative approach to solve the combination sum problem
// - Uses a stack to keep track of the current combination and the remaining target
// - Handles duplicate candidates
vector<vector<int>> combinationSum5(vector<int>& candidates, int target) {
    vector<vector<int>> result;
    sort(candidates.begin(), candidates.end()); // Sort candidates to handle duplicates
    vector<pair<vector<int>, int>> stack; // Store combination and remaining target
    stack.push_back({{}, target});

    while (!stack.empty()) {
        vector<int> combination = stack.back().first;
        int remainingTarget = stack.back().second;
        stack.pop_back();

        if (remainingTarget == 0) {
            result.push_back(combination);
            continue;
        }

        for (int i = (combination.empty() ? 0 : (lower_bound(candidates.begin(), candidates.end(), combination.back()) - candidates.begin())); i < candidates.size(); ++i) {
             if (i > (combination.empty() ? 0 : (lower_bound(candidates.begin(), candidates.end(), combination.back()) - candidates.begin())) && candidates[i] == candidates[i - 1]) {
                continue; // Skip duplicates
            }
            if (remainingTarget - candidates[i] >= 0) {
                vector<int> nextCombination = combination;
                nextCombination.push_back(candidates[i]);
                stack.push_back({nextCombination, remainingTarget - candidates[i]});
            }
        }
    }
    return result;
}

int main() {
    vector<int> candidates = {2, 3, 6, 7};
    int target = 7;
    vector<vector<int>> result1, result2, result3;
    set<vector<int>> result_set;
    vector<int> combination;

    cout << "Candidates: ";
    for (int num : candidates) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Target: " << target << endl << endl;

    // Approach 1: Backtracking with Duplicates Allowed
    combinationSum1(candidates, target, result1, combination, 0);
    cout << "Approach 1: Backtracking with Duplicates Allowed" << endl;
    for (const auto& comb : result1) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Approach 2: Backtracking with Duplicates Allowed (Optimized)
    combination.clear();
    combinationSum2(candidates, target, result2, combination, 0);
    cout << "Approach 2: Backtracking with Duplicates Allowed (Optimized)" << endl;
    for (const auto& comb : result2) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Test Case for Approach 3: Duplicates Not Allowed
    vector<int> candidates3 = {10, 1, 2, 7, 6, 1, 5};
    int target3 = 8;
    vector<vector<int>> result3_1;
    vector<int> combination3;
    sort(candidates3.begin(), candidates3.end()); // Sort for approach 3
    combinationSum3(candidates3, target3, result3_1, combination3, 0);
    cout << "Approach 3: Backtracking with Duplicates Not Allowed" << endl;
    cout << "Candidates: ";
    for(int c : candidates3)
        cout<< c << " ";
    cout << endl;
    cout << "Target: " << target3 << endl;
    for (const auto& comb : result3_1) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    // Approach 4: Using set to avoid Duplicates
    combination.clear();
    combinationSum4(candidates, target, result_set, combination, 0);
    cout << "Approach 4: Using set to avoid Duplicates" << endl;
    for (const auto& comb : result_set) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

     // Approach 5: Iterative Approach using Stack
    vector<vector<int>> result5 = combinationSum5(candidates, target);
    cout << "Approach 5: Iterative Approach using Stack" << endl;
    for (const auto& comb : result5) {
        cout << "[ ";
        for (int num : comb) {
            cout << num << " ";
        }
        cout << "] ";
    }
    cout << endl << endl;

    return 0;
}

File: Java/14-Backtracking/Problems/3-CombinationSum/CombinationSumPro.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum {

    public static void main(String[] args) {
        int[] candidates = {2, 3, 6, 7};
        int target = 7;

        // Approach 1: Basic Backtracking
        System.out.println("Approach 1: Basic Backtracking");
        List<List<Integer>> result1 = combinationSum1(candidates, target);
        printResult(result1);

        // Approach 2: Backtracking with Optimization (Sorting)
        System.out.println("\nApproach 2: Backtracking with Optimization (Sorting)");
        List<List<Integer>> result2 = combinationSum2(candidates, target);
        printResult(result2);

        // Approach 3: Backtracking with Memoization (for distinct combinations - not directly applicable here but shown for concept)
        System.out.println("\nApproach 3: Backtracking with Memoization (Concept - Not directly applicable to original problem, but shown for variation)");
        //  The original problem allows duplicates, so pure memoization doesn't help *directly* with the original problem's output, but it's useful for *counting* distinct combinations.  We'd need a different problem statement (no duplicates allowed in a single combination) to see memoization's full benefit in *returning lists*.  This version shows how memoization *could* be used if the problem were modified to avoid duplicate combinations.
        List<List<Integer>> result3 = combinationSum3(candidates, target);
        printResult(result3);

        // Approach 4: Iterative Approach using a Stack (Simulated Recursion)
        System.out.println("\nApproach 4: Iterative Approach using a Stack (Simulated Recursion)");
        List<List<Integer>> result4 = combinationSum4(candidates, target);
        printResult(result4);

        // Approach 5: Dynamic Programming (Counting Combinations - Variation)
        System.out.println("\nApproach 5: Dynamic Programming (Counting Combinations - Variation)");
        int count = combinationSum5(candidates, target);
        System.out.println("Number of combinations: " + count); // Prints an integer, not a list of lists.
        //  Note:  DP is most natural for *counting* combinations, *not* for generating the lists of combinations when duplicates are allowed.  If the problem were to ask *how many* unique combinations sum to the target, DP would be very efficient.  For the original problem (returning lists with duplicates), backtracking is generally preferred.
    }

    // Helper method to print the result
    private static void printResult(List<List<Integer>> result) {
        for (List<Integer> combination : result) {
            System.out.println(combination);
        }
    }

    // Approach 1: Basic Backtracking
    // Time Complexity: O(2^n) in the worst case, where n is the number of candidates.
    // Space Complexity: O(n) for the depth of the recursion tree.
    public static List<List<Integer>> combinationSum1(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> currentCombination = new ArrayList<>();
        findCombinations1(candidates, target, 0, currentCombination, result);
        return result;
    }

    private static void findCombinations1(int[] candidates, int target, int start, List<Integer> currentCombination, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(currentCombination));
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            currentCombination.add(candidates[i]);
            findCombinations1(candidates, target - candidates[i], i, currentCombination, result);
            currentCombination.remove(currentCombination.size() - 1);
        }
    }

    // Approach 2: Backtracking with Optimization (Sorting)
    // Time Complexity: O(2^n) in the worst case, but often faster than Approach 1 due to pruning.
    // Space Complexity: O(n) for the depth of the recursion tree.
    public static List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> currentCombination = new ArrayList<>();
        Arrays.sort(candidates); // Sort the candidates to optimize (prune unnecessary branches)
        findCombinations2(candidates, target, 0, currentCombination, result);
        return result;
    }

    private static void findCombinations2(int[] candidates, int target, int start, List<Integer> currentCombination, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(currentCombination));
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (target - candidates[i] < 0) { // Optimization: If current candidate is too large, skip it and the rest (since sorted)
                break;
            }
            currentCombination.add(candidates[i]);
            findCombinations2(candidates, target - candidates[i], i, currentCombination, result);
            currentCombination.remove(currentCombination.size() - 1);
        }
    }

    // Approach 3: Backtracking with Memoization (for distinct combinations - CONCEPTUAL EXAMPLE)
    // Time Complexity:  O(n * target)  if we were counting *distinct* combinations (with the problem modified).
    // Space Complexity: O(n * target)
    // IMPORTANT:  This approach is shown for the *concept* of memoization.  For the *original* Combination Sum problem (where the same number can be used multiple times), this memoization does NOT directly apply to reduce the number of *lists* returned.  It's more applicable to a variation where you want to count the number of *distinct* combinations.  To make this code directly usable for a variation of Combination Sum, you would need to change the problem statement (e.g., each number can be used at most once in a combination)
    public static List<List<Integer>> combinationSum3(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> currentCombination = new ArrayList<>();
        //  The memoization table stores boolean values, indicating if a certain target can be reached
        //  from a certain starting index.  It's not storing the lists of combinations themselves
        //  in this example, because with duplicates allowed, the lists aren't "distinct" enough
        //  to memoize effectively for the original problem.
        boolean[][] memo = new boolean[candidates.length + 1][target + 1];
        findCombinations3(candidates, target, 0, currentCombination, result, memo);
        return result;
    }

    private static void findCombinations3(int[] candidates, int target, int start, List<Integer> currentCombination, List<List<Integer>> result, boolean[][] memo) {
        if (target == 0) {
            result.add(new ArrayList<>(currentCombination));
            return;
        }
        if (target < 0) {
            return;
        }
        if (memo[start][target]) { // Check memoization table
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            currentCombination.add(candidates[i]);
            findCombinations3(candidates, target - candidates[i], i, currentCombination, result, memo);
            currentCombination.remove(currentCombination.size() - 1);
        }
        memo[start][target] = true; // Store the result in the memoization table
    }

    // Approach 4: Iterative Approach using a Stack (Simulated Recursion)
    // Time Complexity:  Similar to the recursive approaches, O(2^n) in the worst case.
    // Space Complexity: O(n) for the stack.
    public static List<List<Integer>> combinationSum4(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates); // Important for optimization
        // Stack to simulate the recursion
        java.util.Stack<State> stack = new java.util.Stack<>();
        stack.push(new State(0, new ArrayList<>(), target));

        while (!stack.isEmpty()) {
            State currentState = stack.pop();
            int start = currentState.start;
            List<Integer> currentCombination = currentState.combination;
            int currentTarget = currentState.target;

            if (currentTarget == 0) {
                result.add(new ArrayList<>(currentCombination));
                continue;
            }
            if (currentTarget < 0) {
                continue;
            }

            for (int i = start; i < candidates.length; i++) {
                if (currentTarget - candidates[i] < 0) {
                    break; // Optimization from sorted input
                }
                List<Integer> nextCombination = new ArrayList<>(currentCombination);
                nextCombination.add(candidates[i]);
                stack.push(new State(i, nextCombination, currentTarget - candidates[i]));
            }
        }
        return result;
    }

    // Helper class for the stack in the iterative approach
    private static class State {
        int start;
        List<Integer> combination;
        int target;

        public State(int start, List<Integer> combination, int target) {
            this.start = start;
            this.combination = combination;
            this.target = target;
        }
    }

    // Approach 5: Dynamic Programming (Counting Combinations - Variation)
    // Time Complexity: O(n * target), where n is the number of candidates.
    // Space Complexity: O(target)
    // IMPORTANT: This approach solves a *variation* of the problem: counting the *number* of combinations,
    // NOT generating the combinations themselves, and it's most suitable when you want to count
    // distinct combinations.  For the original problem (generating lists with duplicates), backtracking is better.
    public static int combinationSum5(int[] candidates, int target) {
        int[] dp = new int[target + 1];
        dp[0] = 1; // Base case: there is one way to make a sum of 0 (empty combination)

        for (int candidate : candidates) {
            for (int i = candidate; i <= target; i++) {
                dp[i] += dp[i - candidate];
            }
        }
        return dp[target];
    }
}

